ugvc/mrd/featuremap_consensus_utils.py

```diff
@@ -0,0 +1,211 @@
+    rec.qual = qual_agg_func(record_dict[FeatureMapFields.X_QUAL.value])
+    # INFO fields to aggregate
+    # exceptions: X_QUAL
+    rec.info[FeatureMapFields.X_QUAL.value] = list(record_dict[FeatureMapFields.X_QUAL.value])
+    for field in fields_to_collect["numeric_array_fields"]:
+        if field in record_dict:
+            rec.info[field] = list(record_dict[field])
+    for field in fields_to_collect["string_list_fields"]:
+        if field in record_dict:
+            rec.info[field] = "|".join(record_dict[field])
+
+    # FORMAT fields to aggregate
+    record_dict[VAF] = record_dict[FeatureMapFields.FILTERED_COUNT.value] / record_dict[DP]
+    record_dict[GT] = (0, 1)
+    record_dict[AD] = [
+        record_dict[DP] - record_dict[FeatureMapFields.FILTERED_COUNT.value],
+        record_dict[FeatureMapFields.FILTERED_COUNT.value],
+    ]
+    for format_field in format_fields:
+        rec.samples[sample_name][format_field] = record_dict[format_field]
+    # FILTER field
+    if rec.qual < min_qual:
+        curr_filter = "LowQual"
+    elif rec.qual >= min_qual and record_dict[FeatureMapFields.FILTERED_COUNT.value] == 1:
+        curr_filter = "SingleRead"
+    else:
+        curr_filter = "PASS"
+    rec.filter.add(curr_filter)
+    # write to file
+    out_fh.write(rec)
+    return rec
+
+
+def pileup_featuremap(
+    featuremap: str, output_vcf: str, genomic_interval: str = None, min_qual: int = 0, sample_name: str = "SAMPLE"
+):
+    """
+    Pileup featuremap vcf to a reagular vcf, save the mean quality and read count
+
+    Inputs:
+        featuremap (str): The input featuremap vcf file
+        output_vcf (str): The output pileup vcf file
+        genomic_interval (str): The genomic interval to pileup, format: chr:start-end
+        min_qual (int): The minimum quality threshold
+        sample_name (str): The name of the sample (default: "SAMPLE")
+    Output:
+        output_vcf (str): The output vcf file
+    """
+    cons_dict = defaultdict(dict)
+    # process genomic interval input
+    if genomic_interval is None:
+        chrom = None
+        start = None
+        end = None
+    if genomic_interval is not None:
+        genomic_interval_list = genomic_interval.split(":")
+        chrom = genomic_interval_list[0]
+        start = int(genomic_interval_list[1].split("-")[0])
+        end = int(genomic_interval_list[1].split("-")[1])
+
+    # generate a new header
+    orig_header = pysam.VariantFile(featuremap).header
+    header = pysam.VariantHeader()
+    header.add_meta("fileformat", "VCFv4.2")
+    # header.filters.add("PASS", None, None, "Aggregated quality above threshold, more than a single alternative read")
+    header.filters.add(
+        "SingleRead", None, None, "Aggregated quality above threshold, Only a single read agrees with alternative"
+    )
+    header.filters.add("LowQual", None, None, "Aggregated quality below threshold")
+    for field in fields_to_collect["numeric_aggregated_fields"]:
```

Comment:
if "field" is not in "fields_to_collect", do you want it to continue without it with just a warning, or raise an error? In any case please check that "field in orig_header.info" and if False handle accordingly